modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/EvalQueryQuality.java

```diff
@@ -88,27 +91,31 @@ public List<RatedSearchHit> getHitsAndRatings() {
     }
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject(queryId);
-        builder.field(METRIC_SCORE_FIELD.getPreferredName(), this.metricScore);
-        builder.startArray(UNRATED_DOCS_FIELD.getPreferredName());
-        for (DocumentKey key : EvaluationMetric.filterUnratedDocuments(ratedHits)) {
-            builder.startObject();
-            builder.field(RatedDocument.INDEX_FIELD.getPreferredName(), key.index());
-            builder.field(RatedDocument.DOC_ID_FIELD.getPreferredName(), key.docId());
-            builder.endObject();
-        }
-        builder.endArray();
-        builder.startArray(HITS_FIELD.getPreferredName());
-        for (RatedSearchHit hit : ratedHits) {
-            hit.toXContent(builder, params);
-        }
-        builder.endArray();
-        if (optionalMetricDetails != null) {
-            builder.field(METRIC_DETAILS_FIELD.getPreferredName(), optionalMetricDetails);
-        }
-        builder.endObject();
-        return builder;
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params outerParams) {
+        return Iterators.concat(
+            ChunkedToXContentHelper.singleChunk(
+                (builder, params) -> builder.startObject(queryId),
+                (builder, params) -> builder.field(METRIC_SCORE_FIELD.getPreferredName(), this.metricScore),
+                (builder, params) -> builder.startArray(UNRATED_DOCS_FIELD.getPreferredName())
+            ),
+            Iterators.flatMap(
```

Comment:
In general, I'm never sure about this - should we keep everything as lazy as possible (`flatMap`), or do is it better to "pin down" things (e.g. have an external foreach loop that builds all the needed Iterators and then `concat` them)?
This class appears to be immutable, so the result should be the same. What do you think @original-brownbear ? Maybe we could ask @DaveCTurner advice here.

Reply:
Lazy seems best, we want to delay as much work as possible until it's needed, although I don't really understand the alternative you're suggesting.

Reply:
The alternative is what I used (partially) in SearchHit: build an array of Iterators, then flatMap them. There it was kind of needed/the easiest approach given the if(not null) checks.

In this case it would be something like:
```
var unratedDocs = EvaluationMetric.filterUnratedDocuments(ratedHits)).stream().map(
   key -> ChunkedToXContentHelper.singleChunk(
                    (builder, params) -> builder.startObject(),
                    (builder, params) -> builder.field(RatedDocument.INDEX_FIELD.getPreferredName(), key.index()),
                    (builder, params) -> builder.field(RatedDocument.DOC_ID_FIELD.getPreferredName(), key.docId()),
                    (builder, params) -> builder.endObject()
                )
).toList()

return Iterator.concat(
   // other fields
  Iterators.flatMap(unratedDocs.iterator(), Function.identity())
```

Reply:
I.e. processing `ratedHits` "now" (function call time) to produce a list of ToXContent iterators, instead of generating the ToXContent later, when we are iterating.

Reply:
I see, ok. That's not lazy enough IMO, there's no need to allocate a whole other O(N) list here.

A further advantage of having it as a single expression is that there's no confusion about the order in which things get serialized. Once you start introducing variables to hold parts of the output it takes the reader more effort to line everything up.

Reply:
Fair enough.
Maybe we can build a specialized iterator that uses just the list to iterate, "wrapping" it . This should save the O(N) Iterator wrappers of flatMap/concat, so it will not be additional but alternative.

For the second point I agree - at this point, better to have the whole thing put into the array and transform it to a single Iterator when returning (via `flatMap` or else)

Reply:
> This should save the O(N) Iterator wrappers of flatMap/concat

Maybe I'm missing something, but in what sense are these things O(N)? AFAICT they do no more work, and no more allocation, than any alternatives.

Reply:
Yeah, like I commented below, my bad: I looked at concat quickly and thought it was structured in a recursive way (head-tail). There is definitely no O(N) there. Thanks for all the feedback!